### ISSUE TITLE
master performance issues

### ISSUE THREAD

**Issue by xuzijian-new** [locustio/locust#1963]
Hello, Cyberw. Thank you for reading this message in your busy schedule. Recently, I encountered a problem that has been perplexing me for a long time: during locust distributed pressure test, 250 workers were started to pressure the target server, and the result was 100% master CPU, so the master performance bottleneck was encountered. How did you solve this problem? When master mode starts, how many worker connections can it support? Are there alternatives? My ultimate goal is to set up 10W TCP connections to test the target server, which may require more workers. Looking forward to an early reply

python 3.9.4
locust 2.2.0

**Comment by cyberw** [comment]
Hi! Can you share a way to replicate your test? The kind of problem you are having is not something we can debug based on just words :)

Your issue may be related to the master calculating how to distribute the Users across workers.

Maybe have a look at the test cases in https://github.com/locustio/locust/blob/master/benchmarks/dispatch.py

**Comment by xuzijian-new** [comment]
First of all, thank you for answering the question so I can share the test method. Unfortunately, I can't send you the complete test code due to some irresistible factors. This is my test demo based on the core logic. You can try this debugging. Please feel free to communicate if you have any questions.

`
# coding=utf-8
import time
import gevent
import socket
from gevent import Timeout
from gevent.event import AsyncResult
from locust import User, task, between, events

from locust.exception import StopUser


class WebsiteUser(User):
    wait_time = between(0, 0)

    def on_start(self):
        self.tcp = None
        self.asyncresult_map = {}
        try:
            self.tcp = socket.socket()
            start = time.time()
            self.tcp.connect(("127.0.0.1", 1000))  # target server address
            events.request.fire(
                request_type='get',
                name='tcp_connect',
                response_time=(time.time() - start) * 1000,
                response_length=0,
                context=None,
                exception=None,
            )
        except Exception as e:
            events.request.fire(
                request_type='get',
                name='tcp_connect',
                response_time=0,
                response_length=0,
                context=None,
                exception=e,
            )
        else:
            gevent.spawn(self.read_recv)

    def read_recv(self):
        while self.tcp:
            if self.tcp.detach():
                break
            try:
                data = self.tcp.recv(2 ** 16)
                if not data:
                    break
                self.handle_read(data)
            except:
                raise

    def handle_read(self, data):
        # You might have to deal with packages(data) here
        send_data = self.asyncresult_map.pop("tcp_send", None)
        if send_data:
            ar, start = send_data
            events.request.fire(
                request_type='get',
                name='tcp_send',
                response_time=(time.time() - start) * 1000,
                response_length=0,
                context=None,
                exception=None,
            )
            ar.set(True)

    @task(1)
    def index(self):
        if not self.tcp:
            raise StopUser()

        while not self.tcp.detach():
            ar = AsyncResult()
            start = time.time()
            self.asyncresult_map["tcp_send"] = (ar, start)
            self.tcp.send(b"xxxxxx")  # Sends packets to the server
            try:
                ok = ar.get(timeout=30)
            except Timeout:
                events.request.fire(
                    request_type='get',
                    name='tcp_send',
                    response_time=30 * 1000,
                    response_length=0,
                    context=0,
                    exception='Timeout',
                )


if __name__ == "__main__":
    import locust.main
    import sys

    sys.argv.extend([
        "-f", __file__,
        "--web-host", "127.0.0.1", '--web-port', "8089",
        "--autostart",
        "-u", "100",
        "-r", "1",
    ])

    locust.main.main()
`

**Comment by cyberw** [comment]
I’m confused. Are you running distributed or not? I’ve never used AsyncResult so I’m not sure, but maybe the issue is related to that?

**Comment by cyberw** [comment]
Also, why are you calling locust.main() yourself?

**Comment by xuzijian-new** [comment]
I adopted a distributed deployment, deploying the test code to the Centos7 system. AsyncResult is a blocking behavior class in the GEvent library that waits for a response message to be sent. It is used to collect statistics on the time taken by send to RECV. I don't think this is the problem.

Linux Command:
locust -f test_demo.py --master --expect-workers 250
locust -f test_demo.py --worker --master-host=127.0.0.1 (This command was executed 250 times )

**Comment by xuzijian-new** [comment]
> Also, why are you calling locust.main() yourself?

This was because I was lazy and wanted to run test_demo.py directly through PyCharm, rather than using the command line every time. There is no doubt that the actual stress test was initiated from the command line.

**Comment by xuzijian-new** [comment]
I'd also like to add that I started with 10,000 users and 100 rates. Pressure test for more than 5 minutes.

eg: -u 10000 -r 1000

**Comment by cyberw** [comment]
Are you running the workers on the same machine as master?

**Comment by cyberw** [comment]
And if not, how many machines are you using? Don't run more than 1 worker per cpu core.

**Comment by xuzijian-new** [comment]
Good evening, Cyberw. 
> Are you running the workers on the same machine as master?
    No, they are in different machines.
> And if not, how many machines are you using? Don't run more than 1 worker per cpu core.
   I very much agree with your point of view, and I did that.

The master and worker are basically deployed to work on different machines (they communicate and interact with each other in the internal network). According to the deployment mode of one worker per core, I allocate them to different centos operating systems, and the total number of workers in a single operating system does not exceed the total number of cores minus one.
Eg: 16c32G centos 7, I will deploy a maximum of 15 workers. If the master is deployed in 16c32G centos 7, THEN I will allocate 2 cores for the master and the remaining 14 cores for 13 workers or 0 workers at most

**Comment by cyberw** [comment]
Sounds good. Unfortunately I dont have the time or resources to try to replicate your issue. If you can reproduce it with normal HttpUsers and boil it down to a script (reproducing with all workers running on the same machine could be ok if the workers are just sleeping).

**Comment by mboutet** [comment]
10k users with spawn rate of 100 distributed on 250 workers should not be an issue for the master.

Like @cyberw said, if you can reproduce the problem with a simple "hello world"-type test case, it will be easier to troubleshoot. By this, I mean a single http user type that simply sleeps.

**Comment by xuzijian-new** [comment]
> 10k users with spawn rate of 100 distributed on 250 workers should not be an issue for the master.
> 
> Like @cyberw said, if you can reproduce the problem with a simple "hello world"-type test case, it will be easier to troubleshoot. By this, I mean a single http user type that simply sleeps.

hello, @cyberw @mboutet  I believe 10K users with spawn rate of 100 distributed on 250 workers should not be an issue for the master.

I tried to debug this problem today, and I initially located that it might be related to 'events.request.fire()'. Because in distributed mode, each request.fire action will be synchronized to the master, so the master's load will increase as the number of calls to 'events.request.fire()' in the worker increases until 100%. I think this may be a performance issue. Based on this interpretation, can master support multi-opening or use multi-core to satisfy more workers?

**Comment by xuzijian-new** [comment]
> Sounds good (as in: no obvious mistakes). Unfortunately I dont have the time or resources to try to replicate your issue.
> 
> If you can reproduce it with normal HttpUsers and boil it down to a script (reproducing with all workers running on the same machine could be ok if the workers are just sleeping), I or @mboutet might be able to take a look.

I think this code should reproduce the problem well, as long as the number of workers started is large.

# coding=utf-8
import gevent
from locust import HttpUser, task, between, events


class WebsiteUser(HttpUser):
    wait_time = between(0, 0)
    host = 'http://127.0.0.1:8000/'

    def on_start(self):
        pass

    @task(1)
    def index(self):
        events.request.fire(
            request_type='get',
            name='test_master_rps',
            response_time=300,
            response_length=0,
            context=0,
            exception=None,
        )
        # Simulate a 300 millisecond request
        gevent.sleep(0.3)

**Comment by cyberw** [comment]
> 'events.request.fire()'. Because in distributed mode, each request.fire action will be synchronized to the master

Sorry for the late reply, but this is not the case. Requests are aggregated and batched before they are sent to the master, so there shouldnt be high load on master just because you have lots of events. And any handlers you might have registered for the request event are executed locally, on the worker nodes.
